<a href="https://colab.research.google.com/github/SreeLakshmi1242/Projects/blob/main/Linear_Algebra_and_Calculus.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Advanced Certification Program in Computational Data Science
## A program by IISc and TalentSprint
### Mini Project Notebook: Linear Algebra and Calculus

## Problem Statement

 The task is to advise a petroleum company on how to meet the demands of their customers for motor oil, diesel oil and gasoline.

## Learning Objectives

At the end of the experiment, you will be able to

* create arrays and matrices in python
* understand the concepts of linear equations
* solve the system of linear equations

### Data

From a barrel of crude oil, in one day, factory $A$ can produce
* 20 gallons of motor oil,
* 10 gallons of diesel oil, and
* 5 gallons of gasoline

Similarly, factory $B$ can produce
* 4 gallons of motor oil,
* 14 gallons of diesel oil, and
* 5 gallons of gasoline

while factory $C$ can produce
* 4 gallons of motor oil,
* 5 gallons of diesel oil, and
* 12 gallons of gasoline

There is also waste in the form of paraffin, among other things. Factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude, factory $B$ 5 gallons, and factory $C$ 2 gallons.

**Note:** Your conclusion should include a discussion of the nature of the terms *unique*, *no solution*, *overdetermined* and *underdetermined* as they apply in the context of the oil plants.

## Grading = 10 Points

In [ ]:
#libraries
import numpy as np
import sympy as sy

### Create an array

Create an array of size 2x3 with arbitrary values.

In [ ]:
# YOUR CODE HERE
sample_array = np.array([[1,2,3],[4,5,6]])
print(sample_array)

[[1 2 3]
 [4 5 6]]


### Create the system of Linear Equations

Suppose the current daily demand from distributors is 6600 gallons of motor oil, 5100 gallons of diesel oil and 3100 of gasoline.

Set up the system of equations which describes the above situation. Please include the units as well.

Let the number of barrels used by factory $A$, $B$ and $C$ are $x$, $y$ and $z$ respectively.

Then the system of linear equations will be

$$Motor\ oil:\ \ \ 20x + 4y + 4z = 6600$$

$$Diesel\ oil:\ \ \ 10x + 14y + 5z = 5100$$

$$Gasoline:\ \ \ 5x + 5y + 12z = 3100$$

### Solve the system of Linear Equation (2 points)

How many barrels of crude oil each plant should get in order to meet the demand as a group. Remember that we can only provide each plant with an integral number of barrels.

In [ ]:
# YOUR CODE HERE
demand = sy.Matrix([[20,4,4,6600],[10,14,5,5100],[5,5,12,3100]])
demand.rref()

(Matrix([
 [1, 0, 0, 1149/4],
 [0, 1, 0,  515/4],
 [0, 0, 1,     85]]),
 (0, 1, 2))

Thus, system of equations have a unique solution

In [ ]:
def barrels(x,y,z):
  fact_A = 20*x + 10*x + 5*x
  fact_B = 4*y + 14*y + 5*y
  fact_C = 4*z + 5*z + 12*z
  return print("Factory A: ",round(fact_A,0)," barrels"),print("Factory B: ",round(fact_B,0)," barrels"),print("Factory C: ",round(fact_C,0)," barrels")

In [ ]:
x=1149/4
y=515/4
z=85
barrels(x,y,z)

Factory A:  10054.0  barrels
Factory B:  2961.0  barrels
Factory C:  1785  barrels


(None, None, None)

Suppose the total demand for all products **doubled**. What would the solution now be? How does it compare to the original solution? Why, mathematically, should this have been expected?

In [ ]:
# YOUR CODE HERE
double_demand = sy.Matrix([[20,4,4,2*6600],[10,14,5,2*5100],[5,5,12,2*3100]])
double_demand.rref()

(Matrix([
 [1, 0, 0, 1149/2],
 [0, 1, 0,  515/2],
 [0, 0, 1,    170]]),
 (0, 1, 2))

Even when the demand doubled, the system of equations have a unique solution

In [ ]:
x=1149/2
y=515/2
z=170
barrels(x,y,z)

Factory A:  20108.0  barrels
Factory B:  5922.0  barrels
Factory C:  3570  barrels


(None, None, None)

Doubling of demand increased the requirement of barrel by factories twice. This is because, doubling on R.H.S will cause doubling on L.H.S. Distributing this will cause each companies requirement to double.

Suppose that the company acquires another group of distributors and that the daily demand of this group is 2000 gallons of motor oil, 4000 gallons of gasoline, and 4000 gallons of diesel oil. How would you set up production of just this supply? Are there any options (more than one way)?

In [ ]:
# YOUR CODE HERE
new_demand = sy.Matrix([[20,4,4,2000],[10,14,5,4000],[5,5,12,4000]])
print(new_demand)
# another way
new_demand_2 = np.array([[20,4,4,2000],[10,14,5,4000],[5,5,12,4000]])
print(new_demand_2)
new_demand.rref()

Matrix([[20, 4, 4, 2000], [10, 14, 5, 4000], [5, 5, 12, 4000]])
[[  20    4    4 2000]
 [  10   14    5 4000]
 [   5    5   12 4000]]


(Matrix([
 [1, 0, 0,  25/2],
 [0, 1, 0, 375/2],
 [0, 0, 1,   250]]),
 (0, 1, 2))

Next, calculate the needs of each factory (in barrels of crude, as usual) to meet the total demand of both groups of distributors. When you have done this, compare your answer to results already obtained. What mathematical conclusion can you draw?

In [ ]:
# YOUR CODE HERE
comb_demand = sy.Matrix([[20,4,4,2000+6600],[10,14,5,4000+5100],[5,5,12,4000+3100]])
solution = comb_demand.rref()
print(solution)

(Matrix([
[1, 0, 0, 1199/4],
[0, 1, 0, 1265/4],
[0, 0, 1,    335]]), (0, 1, 2))


The system of equations have a unique solution

In [ ]:
x=1199/4
y=1265/4
z=335
barrels(x,y,z)

Factory A:  10491.0  barrels
Factory B:  7274.0  barrels
Factory C:  7035  barrels


(None, None, None)

### Sensitivity and Robustness (1 point)

In real life applications, constants are rarely ever exactly equal to their stated value; certain amounts of uncertainty are always present. This is part of the reason for the science of statistics. In the above model, the daily productions for the plants would be averages over a period of time. Explore what effect small changes in the parameters have on the output.

To do this, pick any 3 coefficients, one at a time, and increase or decrease them by 3%. For each case , note what effect this has on the solution, as a percentage change. Can you draw any overall conclusion?

In [ ]:
# YOUR CODE HERE
for i in range(0,3):
  to_imp = comb_demand.copy()
  to_imp[:,i]=to_imp[:,i]*1.03
  sol = to_imp.rref()
  change=(sol[0][:,3]-solution[0][:,3])
  increment = []
  for i in range(3):
    inc=100*change[i]/solution[0][:,3][i]
    increment.append(round(inc,0))
  print(increment)

[-3.00000000000000, 0.0, 0.0]
[0, -3.00000000000000, 0]
[0, 0, -3.00000000000000]


Increasing the coefficient value impacted the demand for the product in which the coefficient contributed the most.
Thus,Motor oil's demand was impacted by change in x only, Diesel's demand was impacted by change in y only and Gasoline's demand was impacted by change in z only.

### A Plant Off-Line (1 point)

Suppose factory $C$ is shut down by the EPA (Environmental Protection Agency) temporarily for excessive emissions into the atmosphere. If your demand is as it was originally (6600, 5100, 3100), what would you now say about the companies ability to meet it? What do you recommend they schedule for production now?

In [ ]:
# YOUR CODE HERE
shut_demand = sy.Matrix([[20,4,6600],[10,14,5100],[5,5,3100]])
shut_demand.rref()


(Matrix([
 [1, 0, 0],
 [0, 1, 0],
 [0, 0, 1]]),
 (0, 1, 2))

The above is overdetermined system of equations. From the results, we see the system of equations is inconsistent as it has no sulution. Thus,the company is incapable of meeting the demand with the existing plants. There is a requirement for an additional plant to meet demand.


### Buying another plant

####(Note the following given information. You will see questions in continuation to this, in the subsequent sections)

This situation has caused enough concern that the CEO is considering buying another plant, identical to the third, and using it permanently. Assuming that all 4 plants are on line, what production do you recommend to meet the current demand (5000, 8500, 10000)? In general, what can you say about any increased flexibility that the 4th plant might provide?

Let the number of barrels used by factory $A$, $B$, $C$ and $D$ are $x$, $y$, $z$ and $w$ respectively.

Then the system of linear equations will be

$$20x + 4y + 4z + 4w = 5000$$

$$10x + 14y + 5z + 5w = 8500$$

$$5x + 5y + 12z + 12w = 10000$$

The above system of linear equation has fewer equations than variables, hence it is *underdetermined* and cannot have a unique solution. In this case, there are either infinitely many solutions or no exact solution. We can solve it by keeping $w$ as constant and using [rref](http://linear.ups.edu/html/section-RREF.html) form to solve the system of linear equation.

To know about rref implementation in python refer [here](https://docs.sympy.org/latest/tutorial/matrices.html#rref).

In [ ]:
import sympy as sy

# create symbol 'w'
w = sy.Symbol("w")
A_aug = sy.Matrix([[20, 4, 4, 5000-4*w],
                   [10, 14, 5, 8500-5*w],
                   [5, 5, 12, 10000-12*w]])
# show rref form
A_aug.rref()

(Matrix([
 [1, 0, 0,   195/4],
 [0, 1, 0,  1325/4],
 [0, 0, 1, 675 - w]]),
 (0, 1, 2))

From the above result, it can be seen that 4th plant will share the number of barrels required by the 3rd plant only, while the requirement of 1st and 2nd plant will remain unaffected.

### Calculate the amount of Paraffin supplied (1 point)

The company has just found a candle company that will buy its paraffin. Under the current conditions (i.e, after buying another plant) for demand (5000, 8500, 10000), how much can be supplied to them per day?

According to the problem statement, factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude oil, factory $B$ 5 gallons, and factory $C$ 2 gallons.

In [ ]:
# YOUR CODE HERE
p_a = 3
p_b = 5
p_c = 2
p_d = 2
x = 195/4
y=1325/4
z = 675-w
# fact_A = 20*x + 10*x + 5*x
# fact_B = 4*y + 14*y + 5*y
# fact_C = 4*z + 5*z + 12*z
# fact_D = 4*w + 5*w + 12*w
p_supply = p_a*x + p_b*y + p_c*z + p_d*w
print("Paraffin that can be supplied per day is: ",p_supply,'gallons')

Paraffin that can be supplied per day is:  3152.50000000000 gallons


### Selling the first plant (1 point)

The management is also considering selling the first plant due to aging equipment and high workman's compensation costs for the state it is located in. They would like to know what this would do to their production capability. Specifically, they would like an example of a demand they could not meet with only plants 2 and 3, and also what effect having plant 4 has (recall it is identical to plant 3). They would also like an example of a demand that they could meet with just plants 2 and 3. Any general statements you could make here would be helpful.

Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

When considering only plants 2 and 3, and demand (5000, 8500, 10000) then we have

$$4y + 4z = 5000$$

$$14y + 5z = 8500$$

$$5y + 12z = 10000$$

In [ ]:
# YOUR CODE HERE
sell_d = sy.Matrix([[4,4,5000],[14,5,8500],[5,12,10000]])
sell_d.rref()

(Matrix([
 [1, 0, 0],
 [0, 1, 0],
 [0, 0, 1]]),
 (0, 1, 2))

The system of equations are over determined. Thus, there is no solution. Therefore, the company cannot meet demand with only 2 plants

Taking 4th plant into consideration.
Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

Then for demand (5000, 8500, 10000) the system of linear equations will be

$$4y + 4z + 4w = 5000$$

$$14y + 5z + 5w = 8500$$

$$5y + 12z + 12w = 10000$$

Solve it using rref form.

In [ ]:
# YOUR CODE HERE
sell_d_2 = sy.Matrix([[4,4,4,5000],[14,5,5,8500],[5,12,12,10000]])
sell_d_2.rref()

(Matrix([
 [1, 0, 0, 0],
 [0, 1, 1, 0],
 [0, 0, 0, 1]]),
 (0, 1, 3))

From the row reduced echelon form, the system of equations are inconsistent.Therefore no solution. This is because plant C and D are identical.
Thus, company cannot meet demand with plants B, C and D.

Now, changing demand to (6600, 5100, 3100) and solving the system of equation using rref form.

In [ ]:
# YOUR CODE HERE
sell_d_3 = sy.Matrix([[4,4,4,6600],[14,5,5,5100],[5,12,12,3100]])
sell_d_3.rref()

(Matrix([
 [1, 0, 0, 0],
 [0, 1, 1, 0],
 [0, 0, 0, 1]]),
 (0, 1, 3))

Change in demand will bring no change in solution as the rows are still identical. Thus, company cannot meet demand with plants B, C and D even if demand changes

### Set rates for Products (1 point)

Company wants to set the rates of motor oil, diesel oil, and gasoline. For this purpose they have few suggestions given as follows:

* 100, 66, 102 Rupees per gallon,

* 104, 64, 100 Rupees per gallon,

* 102, 68, 98 Rupees per gallon, and

* 96, 68, 100 Rupees per gallon

for motor oil, diesel oil, and gasoline respectively.

Using matrix multiplication, find the rates which result in maximum total price.

Let $M$ denote the matrix such that rows represents different plants (A, B and C), columns represents different products (motor oil, diesel oil and gasoline) and each value represents production of that product from one barrel of crude oil for that plant.

$$M = \begin{bmatrix}
20 & 10 & 5 \\
4 & 14 & 5  \\
 4 & 5 & 12  
\end{bmatrix}$$

Also, $R$ is a matrix having different rates as its columns.

$$R = \begin{bmatrix}
100 & 104 & 102 & 96 \\
66 & 64 & 68 & 68  \\
102 & 100 & 98 & 100  
\end{bmatrix}$$

In [ ]:
# YOUR CODE HERE
M = np.array([[20,10,5],[4,14,5],[4,5,12]])
R = np.array([[100,104,102,96],[66,64,68,68],[102,100,98,100]])
t_p = np.dot(M,R)
max_p = np.sum(t_p, axis=0)
max_index = np.argmax(max_p)
max_col = R[:,max_index]
print("Max total price : ",max_p[max_index],'Rupees',"\nSuggested rate :",max_col)

Max total price :  6984 Rupees 
Suggested rate : [102  68  98]


Thus, the maximum total price is 6984 Rupees. Thus,  102, 68, 98 Rupees per gallon for motor oil, diesel oil, and gasoline respectively can earn them the maximum

### Marginal Cost (1 point)

The total cost $C(x)$ in Rupees, associated with the production of $x$ gallons of gasoline is given by

$$C(x) = 0.005 x^3 – 0.02 x^2 + 30x + 5000$$

Find the marginal cost when $22$ gallons are produced, where, marginal cost means the instantaneous rate of change of total cost at any level of output.

In [ ]:
# YOUR CODE HERE
x = sy.Symbol("x")
c_x = 0.005*x**3 - 0.02*x**2 + 30*x + 5000
cost_diff=sy.Derivative(c_x,x).doit()
print(cost_diff)
MC=cost_diff.subs(x,22)
print("Marginal Cost when 22 gallons are produced is :",round(MC,2),"Rupees")

0.015*x**2 - 0.04*x + 30
Marginal Cost when 22 gallons are produced is : 36.38 Rupees


### Marginal Revenue (1 point)

The total revenue in Rupees received from the sale of $x$ gallons of a motor oil is given by $$R(x) = 3x^2 + 36x + 5.$$

Find the marginal revenue, when $x = 28$, where, marginal revenue means the rate of change of total revenue with respect to the number of items sold at an instant.

In [ ]:
# YOUR CODE HERE
r_x = 3*x**2 + 36*x + 5
rev_diff=sy.Derivative(r_x,x).doit()
print(rev_diff)
MC=rev_diff.subs(x,28)
print("Marginal Cost when 22 gallons are produced is :",round(MC,2),"Rupees")

6*x + 36
Marginal Cost when 22 gallons are produced is : 204 Rupees


### Pouring crude oil in tank (1 point)

In a cylindrical tank of radius 10 meter, crude oil is being poured at the rate of 314 cubic meter per hour. Then find

* the rate at which the height of crude oil is increasing in the tank, and
* the height of crude oil in tank after 2 hours.

In [ ]:
# YOUR CODE HERE
r=10
dv_dt = 314
dh_dt=dv_dt/(np.pi * r**2)
print("Rate at which height increase is : ", round(dh_dt,0),"meter per hour")
t=2
height_after_2hr = dh_dt*t
print("Height after 2 hours is : ", round(height_after_2hr,0),"meter")





Rate at which height increase is :  1.0 meter per hour
Height after 2 hours is :  2.0 meter
